Editorial review: Document popover="hint" (#37990)
mdn/content: files/en-us/web/api/popover_api/using/index.md

```js
const tooltips = document.querySelectorAll(".tooltip");
const btns = document.querySelectorAll("#button-bar button");
```

Next, we create a function, `addEventListeners()`, which sets four event listeners (via {{domxref("EventTarget.addEventListener()")}}) on a given {{htmlelement("button")}}, chosen by grabbing the `<button>` at a specific index value of the `btns` `NodeList`. The functions act on the `hint` popover at the same index value of the `tooltips` `NodeList`, allowing us to keep the buttons and the tooltips in sync — showing/hiding the correct tooltip when a button is interacted with.

The event listeners [show](/en-US/docs/Web/API/HTMLElement/showPopover) the popover on [`mouseover`](/en-US/docs/Web/API/Element/mouseover_event) and [`focus`](/en-US/docs/Web/API/Element/focus_event), and [hide](/en-US/docs/Web/API/HTMLElement/hidePopover) the popover on [`mouseout`](/en-US/docs/Web/API/Element/mouseout_event) and [`blur`](/en-US/docs/Web/API/Element/blur_event), meaning that the tooltips can be accessed via mouse and keyboard.

```js
function addEventListeners(i) {
  btns[i].addEventListener("mouseover", () => {

>     tooltips[i].showPopover();

Reviewer: See my comment on the demos PR, but ideally these would be `showPopover({source: btns[i]})`.

See our [Nested popover menu example](https://mdn.github.io/dom-examples/popover-api/nested-popovers/) ([source](https://github.com/mdn/dom-examples/tree/main/popover-api/nested-popovers)) for an example. You'll notice that quite a few event handlers have been used to display and hide the subpopover appropriately during mouse and keyboard access, and also to hide both menus when an option is selected from either. Depending on how you handle loading of new content, either in an SPA or multi-page website, some of all of these may not be necessary, but they have been included in this demo for illustrative purposes.

## Using "hint" popover state

There is a third type of popover you can create — **hint popovers**, designated by setting `popover="hint"` on your popover element. `hint` popovers are similar to `auto` popovers, but with a significant difference. They can be light-dismissed, but they do not light-dismiss `auto` popovers when shown, only `hint` popovers.

> This is useful for situations where, for example, you have toolbar buttons that can be pressed to show UI popovers, but you also want to reveal tooltips when the buttons are hovered, without dismissing the UI popovers.

Reviewer 2: A few thoughts 

- The example cannot demonstrate that hint popovers can be light dismissed. The example seems to indicate that additionally when an auto popover is light dismissed popovers the hint is not.
- Can these be used with manual popovers - should they be? Nowhere in the doc do we say.
- I don't love this sentence, partially because of the use of light-dismiss when you mean close. Same comments/reasons as https://github.com/mdn/content/pull/37990/files#r1955356883

> There is a third type of popover you can create — **hint popovers**, designated by setting `popover="hint"` on your popover element. 
> These can be displayed over `auto` popovers without causing them to close, and enable use cases such as tooltips over nested menus. Note that `hint` popovers will close other hint popovers when they are displayed, and can be light-dismissed.

Reviewer 2: PS Very much like the rest of this example.

Author: > The example cannot demonstrate that hint popovers can be light dismissed. The example seems to indicate that additionally when an auto popover is light dismissed popovers the hint is not.

I'm not really sure what to make of this comment. You can quite easily check that a hint popover can be light dismissed by hovering over a button and then pressing Esc. When you light dismiss, whatever popover is focussed at the time is the one that is dismissed. I'm not sure this needs any kind of change.

> Can these be used with manual popovers - should they be? Nowhere in the doc do we say.

Yes, they can, but there is not much point. I've added a note to address this.

> I don't love this sentence, partially because of the use of light-dismiss when you mean close.

OK; I've swapped "dismissed" for "closed" in that sentence.

See our [Nested popover menu example](https://mdn.github.io/dom-examples/popover-api/nested-popovers/) ([source](https://github.com/mdn/dom-examples/tree/main/popover-api/nested-popovers)) for an example. You'll notice that quite a few event handlers have been used to display and hide the subpopover appropriately during mouse and keyboard access, and also to hide both menus when an option is selected from either. Depending on how you handle loading of new content, either in an SPA or multi-page website, some of all of these may not be necessary, but they have been included in this demo for illustrative purposes.

> ## Using "hint" popover state

Reviewer 3: I'm not sure what the policy is on documenting stuff that only applies in the middle stage where some browsers support hint but others don't. But it might be worth pointing out that popover="hint" will fallback to a manual popover in unsupporting browsers rather than auto.

Author: I've added a quick note in my next commit. I think it's worth including, and we can always remove it when all modern browsers catch up.

```js
function addEventListeners(i) {
  btns[i].addEventListener("mouseover", () => {
    tooltips[i].showPopover({ source: btns[i] });
  });

  btns[i].addEventListener("mouseout", () => {
    tooltips[i].hidePopover();
  });

  btns[i].addEventListener("focus", () => {
    tooltips[i].showPopover({ source: btns[i] });
  });

  btns[i].addEventListener("blur", () => {
    tooltips[i].hidePopover();
  });
}
```

Finally, we use a [`for`](/en-US/docs/Web/JavaScript/Reference/Statements/for) loop to iterate through the `<buttons>` in the `btns` `NodeList`, calling the `addEventListeners()` function for each one so that all of them have the desired event listeners set.

```js
for (let i = 0; i < btns.length; i++) {
  addEventListeners(i);
}
```

## Styling popovers

> The Popover API has some related CSS techniques available that are worth looking at.

Reviewer 2: If these are part of the same spec, which I believe they are, this might be better as.

```suggestion
Popovers can be styled and positioned using CSS.
```

Author: You are not wrong, but "Popovers can be styled and positioned using CSS." sounds a bit obvious — popovers are HTML elements, so of course they can be styled using CSS!

I thought about it a bit, and ended up going with:

> This section covers some CSS selection and positioning techniques relevant to popovers.

### Selecting popovers

You can select all popovers with a simple attribute selector:

```css
[popover] {
  /* Declarations here */
}
```

Alternatively, you can select a specific popover type by including a value in the attribute selector:

```css
[popover="auto"] {
  /* Declarations here */
}
```

> You can select only popovers that are showing using the {{cssxref(":popover-open")}} pseudo-class:

Reviewer 2: Perhaps (i.e. I am not sure why "showing" doesn't work as well for me.)

```suggestion
You can select only popovers that are being displayed using the {{cssxref(":popover-open")}} pseudo-class:
```

```css
:popover-open {
  /* Declarations here */
}
```

### Styling the popover backdrop

> The {{cssxref("::backdrop")}} pseudo-element is a full-screen element placed directly behind showing popover elements in the {{glossary("top layer")}}, allowing effects to be added to the page content behind the popover(s) if desired. You might for example want to blur out the content behind the popover to help focus the user's attention on it:

Reviewer 2: Again, I don't like "showing" as an adjective. Consider

```suggestion
The {{cssxref("::backdrop")}} pseudo-element is a full-screen element placed directly behind open popover elements in the {{glossary("top layer")}}, allowing effects to be added to the page content behind the popover(s) if desired. You might for example want to blur out the content behind the popover to help focus the user's attention on it:
```

Author: Again, "showing" is the correct semantic term here. "Open" isn't correct.

> 

Reviewer 2: I can't link above to top of this example. But ...

See the first sentence above:

> There is a third type of popover you can create — **hint popovers**, designated by setting `popover="hint"` on your popover element. `hint` popovers are similar to `auto` popovers, but with a significant difference. They can be light-dismissed, but they do not light-dismiss `auto` popovers when shown, only `hint` popovers.

- The example cannot demonstrate that hint popovers can be light dismissed. The example seems to indicate that additionally when an auto popover is light dismissed popovers the hint is not.
- Can these be used with manual popovers - should they be? Nowhere in the doc do we say.
- I don't love this sentence. Would do something like:

> There is a third type of popover you can create — **hint popovers**, designated by setting `popover="hint"` on your popover element. 
> These can be displayed over `auto` popovers without causing them to close/light dismiss, and enable use cases such as tooltips over nested menus. Note that `hint` popovers can be light-dismissed, and will close other hint popovers when they are displayed.

See our [Popover blur background example](https://mdn.github.io/dom-examples/popover-api/blur-background/) ([source](https://github.com/mdn/dom-examples/tree/main/popover-api/blur-background)) for an idea of how this renders.

### Implicit popover anchor associations

Associating a popover with a control using the [`popovertarget`](/en-US/docs/Web/HTML/Element/button#popovertarget) and [`id`](/en-US/docs/Web/HTML/Global_attributes/id) attributes creates an implicit anchor reference between the two, as does programmatically associating a popover action such as {{domxref("HTMLElement.showPopover", "showPopover()")}} with a control using the `source` option.

This makes it very convenient to position popovers relative to their controls using [CSS anchor positioning](/en-US/docs/Web/CSS/CSS_anchor_positioning). Explicit associations do not need to be made using the {{cssxref("anchor-name")}} and {{cssxref("position-anchor")}} properties.

> For an example that uses this implicit association, see the [Using "hint" popover state](/en-US/docs/Web/API/Popover_API/Using#using_hint_popover_state) section above.

Reviewer 2: Just to be clear, you're saying that the implicit positioning is already in place, so you don't need to worry about it - a hint will appear in the right place above its associated element?

The first sentence doesn't quite say that "This makes it very convenient" - i.e. it sounds like something you have to do, and that you ware about to describe.

The second sentence says "for an example that uses this implicit association see ... you might say "You can see this implicit associate at work in the exmaple XXX: no special CSS had to be created to position the hints near their associated elements.

Author: Only the association is implicit, not the positioning. The implicit association means you don't need to create an explicit anchor reference, for example using the `anchor-name` and `position-anchor` properties.

You still need to do the anchor positioning explicitly, for example, using the `anchor()` function as values of inset properties, or the `position-area` property.

I've not made much of a change here, except that I've added a link to the relevant section of the main anchor positioning guide that explains this association, and how the actual positioning is a separate step.

I've also added this link to the other pages that mention the implicit association, as I thought it would be helpful to clear up confusion.

Let me know if you think this needs anything else.

Reviewer 2: @chrisdavidmills Thanks for explaining - I get it now, but I can completely see why I made this assumption - you show no CSS.

The linked section in "For an example that uses this implicit association, see the [Using "hint" popover state](/en-US/docs/Web/API/Popover_API/Using#using_hint_popover_state) section above." does not provide any example. Nor does the rest of the code. Everything says there are implicit links and nothing shows how you style them.

I don't think it is enough to link to the generic styling section. We need a real example. Something like this, copied from the example source code, with explanation:

```css
[popover="hint"] {
  inset: unset;
  position: absolute;
  top: calc(anchor(bottom) + 5px);
  justify-self: anchor-center;
  margin: 0;

  padding: 8px;
  border-radius: 6px;
  background: #271717;
  color: whitesmoke;
  box-shadow: 1px 1px 3px #999;
  border: none;
}
```

I would also rename https://pr37990.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/Popover_API/Using#implicit_popover_anchor_associations to "Styling hints" - but that's up to you.

Reviewer 3: I don't think styling hints makes sense as a section name fwiw, because you get implicit anchoring for auto and manual popovers too.

Author: OK, I understand the problem now! ;-)

So, in my next commit, I have overhauled this section, explaining that you still need to add the positioning CSS code, and showing examples of what this looks like. 

I have also removed the link to the earlier section about `popover="hint"`, and just linked straight to the code example on GitHub for an actual real example, so they can just jump into the code and check it out themselves. 

I think it is better to not link between these two sections, as it creates a weird circular dependency that is not really needed.

Author: I have also added a new section earlier on in the article called "Popover accessibility features", which explains how popover/invoker relationships result in useful ARIA relationships and focus order updates.

Reviewer 3: Just to be clear you don't get the aria when the invokers are done via JS because it can be any element and we don't know that it makes sense for a given element. But you *should* get the focus fixup.

> ### Implicit popover anchor associations

Reviewer 2: FWIW I still think I'm more likely to find this topic if you call it "Positioning a popover" or "Position a popover relative to its invoker". This heading only means something to me because I already understand it.

Author: I'm going to push back on this. I don't think changing the heading to something like this would make it particularly easier to find the section, and it would also be kind of inaccurate as to the focus of the section. It's not about positioning the popovers; that's covered in the previous section. This section is specifically about the fact that you don't need to include the anchor position association CSS features the elements in question are popovers.

Author: As a compromise, I have promoted it to an h2, to make it more visible on the actual page.

Reviewer 2: Respectfully, disagree. As a reader I don't care at all about implicit anchor stuff at all - that's just glue to what I want to achieve - I can't guess what it does from the name.  What I care about about is what the feature does for me - which is "Positioning the popover relative to its invoker".
Probably this point made more sense when it was an H3 inside the styling section though :-)

Given that this might not convince you, perhaps another compromise would be to update the Styling section. 

Right now it says "To override the default styles and get the popover to appear somewhere else on your viewport, you would need to override the above styles with something like this:". That kind of implies that "this is the only way that you can style/position" but we have another really cool way ...

So a forward reference to your implicit anchor section stating that "the implicit anchor section that follows shows how you can position your popover relative to its invoker rather than relative to the view" (or similar).

Hope that makes sense.

Author: @hamishwillee darn it, I think you are right, having thought about this some more. Although I'd never admit that to anyone ;-)

I have done a whole bunch of restructuring in my latest commit and brought the "Implicit popover anchor associations" section back under the "Styling popovers" with a new subheading of "Popover anchor positioning". I've also updated all the links to that section to suit.

I'm pretty pleased with this new structure, and I think you will be, too.

Reviewer 2: Yes, this version is a thing of beauty. Thank you!!!!!

It is always much easier to be critical of other's work than to come up with something good.

> Associating a popover with its invoker using the [`popovertarget`](/en-US/docs/Web/HTML/Element/button#popovertarget) attribute or the `source` option of the {{domxref("HTMLElement.showPopover()")}} or {{domxref("HTMLElement.togglePopover()")}} methods creates an implicit anchor reference between the two.

Reviewer 2: **Might** be worth spelling out that this works for both  hints and autos.

Reviewer 2: PS I very much like this update. I hope you agree it was worth the effort :-)

Author: > **Might** be worth spelling out that this works for both hints and autos.

It works for manual popovers too. I've clarified this by changing the first line to "Associating any kind of popover..."

> PS I very much like this update. I hope you agree it was worth the effort :-)

Indeed, I do ;-)
